modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/MappingServiceImpl.java

```diff
@@ -220,7 +231,9 @@ private CompletableFuture<MappedFragments> mapFragments(
         return res.thenApply(assignments -> {
             Int2ObjectMap<ExecutionTarget> targetsById = new Int2ObjectOpenHashMap<>();
 
-            MappingContext context = new MappingContext(localNodeName, new ArrayList<>(assignments.nodes()), template.cluster);
+            MappingContext context = new MappingContext(
+                    localNodeName, new ArrayList<>(assignments.nodes(nodeExclusionFilter)), template.cluster
```

Comment:
Do we need a copy to ArrayList here if filter is null?

Reply:
If we can't avoid copying, then the filter can be not null (@Nullable not needed) and null value replaced with "always false" predicate.

Reply:
Change {{DistributionHolder}} to return list instead of set. Now the collection is copied only once when filter is null, otherwise we create filtered list.